staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go

```diff
@@ -0,0 +1,304 @@
+	var tm metav1.TypeMeta
+	if err := laxDecMode.Unmarshal(data, &tm); err != nil {
+		return nil, fmt.Errorf("unable to determine group/version/kind: %w", err)
+	}
+	actual := tm.GetObjectKind().GroupVersionKind()
+	return &actual, nil
+}
+
+type Serializer interface {
+	runtime.Serializer
+	recognizer.RecognizingDecoder
+}
+
+var _ Serializer = &serializer{}
+
+type serializer struct {
+	metaFactory metaFactory
+	creater     runtime.ObjectCreater
+	typer       runtime.ObjectTyper
+}
+
+func NewSerializer(creater runtime.ObjectCreater, typer runtime.ObjectTyper) Serializer {
+	return newSerializer(&defaultMetaFactory{}, creater, typer)
+}
+
+func newSerializer(metaFactory metaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper) *serializer {
+	return &serializer{
+		metaFactory: metaFactory,
+		creater:     creater,
+		typer:       typer,
+	}
+}
+
+func (s *serializer) Identifier() runtime.Identifier {
+	return "cbor"
+}
+
+func (s *serializer) Encode(obj runtime.Object, w io.Writer) error {
+	// https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor
+	if _, err := w.Write([]byte("\xd9\xd9\xf7")); err != nil {
+		return err
+	}
+
+	e := encMode.NewEncoder(w)
+	if u, ok := obj.(runtime.Unstructured); ok {
+		return e.Encode(u.UnstructuredContent())
+	}
+	return e.Encode(obj)
+}
+
+// gvkWithDefaults returns group kind and version defaulting from provided default
+func gvkWithDefaults(actual, defaultGVK schema.GroupVersionKind) schema.GroupVersionKind {
+	if len(actual.Kind) == 0 {
+		actual.Kind = defaultGVK.Kind
+	}
+	if len(actual.Version) == 0 && len(actual.Group) == 0 {
+		actual.Group = defaultGVK.Group
+		actual.Version = defaultGVK.Version
+	}
+	if len(actual.Version) == 0 && actual.Group == defaultGVK.Group {
+		actual.Version = defaultGVK.Version
+	}
+	return actual
+}
+
+// diagnose returns the diagnostic encoding of a well-formed CBOR data item.
+func diagnose(data []byte) string {
+	diag, err := diagMode.Diagnose(data)
+	if err != nil {
+		// Since the input must already be well-formed CBOR, converting it to diagnostic
+		// notation should not fail.
+		util.HandleError(err)
+
+		return hex.EncodeToString(data)
+	}
+	return diag
+}
+
+func (s *serializer) Decode(data []byte, gvk *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	actual, err := s.metaFactory.Interpret(data)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if gvk != nil {
+		*actual = gvkWithDefaults(*actual, *gvk)
+	}
+
+	if into != nil {
+		// TODO: The GVK extracted from the initial pass over the input is only used in for
+		// the "into Unstructured" case if the full decode fails. Consider moving this block
+		// to the top.
+		if u, ok := into.(runtime.Unstructured); ok {
+			dst := map[string]interface{}{}
+			if err := decMode.Unmarshal(data, &dst); err != nil {
+				return nil, actual, err
+			}
+			// TODO: The UnstructuredList implementation of SetUnstructuredContent is
```

Comment:
Once the k/k issue for "this is what's left for CBOR" is present, I want to add these TODOs to the list prior to merge.